src/Features/Core/Portable/CodeLens/CodeLensFindReferenceProgress.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeLens
+{
+    /// <summary>
+    /// Tracks incremental progress of a find references search, we use this to
+    /// count the number of references up until a certain cap is reached and cancel the search
+    /// or until the search completes, if such a cap is not reached.
+    /// </summary>
+    /// <remarks>
+    /// All public methods of this type could be called from multiple threads.
+    /// </remarks>
+    internal sealed class CodeLensFindReferencesProgress : IFindReferencesProgress, IDisposable
+    {
+        /// <summary>
+        /// this token is linked to an aggregate token that is passed to the Find References
+        /// operation, this is used solely to trigger the cancellation of an in progress
+        /// find references operation, when the references count hits a given cap.
+        /// </summary>
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private readonly CancellationTokenSource _aggregateCancellationTokenSource;
+        private readonly SyntaxNode _queriedNode;
+        private readonly ISymbol _queriedSymbol;
+        private readonly ConcurrentSet<Location> _locations;
+
+        /// <remarks>
+        /// If the cap is 0, then there is no cap.
+        /// </remarks>
+        public int SearchCap { get; }
+
+        /// <summary>
+        /// The cancellation token that aggregates the original cancellation token + this progress
+        /// </summary>
+        public CancellationToken CancellationToken => _aggregateCancellationTokenSource.Token;
+
+        public bool SearchCapReached => SearchCap != 0 && ReferencesCount > SearchCap;
+
+        public int ReferencesCount => _locations.Count;
+
+        public ImmutableArray<Location> Locations => _locations.ToImmutableArray();
+
+        public CodeLensFindReferencesProgress(
+            ISymbol queriedDefinition,
+            SyntaxNode queriedNode,
+            int searchCap,
+            CancellationToken cancellationToken)
+        {
+            _queriedSymbol = queriedDefinition;
+            _queriedNode = queriedNode;
+            _cancellationTokenSource = new CancellationTokenSource();
+            _aggregateCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(
```

Comment:
aggregateCancellationTokenSource here can be created from just one cancellationToken.